tests/RobotFramework/tests/tedge_watchdog/tedge_watchdog.robot

```diff
@@ -0,0 +1,135 @@
+
+Suite Setup         Custom Setup
+Suite Teardown      Custom Teardown
+
+Test Tags           adapter:docker    theme:tedge-watchdog
+
+
+*** Variables ***
+# Interval at which we send notifications/check health of the service. Use smaller value for test to complete faster but
+# increase if flaky or debugging.
+${WATCHDOG_SEC}     60
+${SERVICE_NAME}     tedge-agent
+
+
+*** Test Cases ***
+# NOTE: This test succeeds, but not for the reason we expect - instead of simply not sending a notification, upon
+# reciving a status:down message tedge-watchdog exits for unknown reason, so it can't send the notification and the
+# monitored service is restarted by systemd, along with tedge-watchdog itself.
+# This is a result of a specific configuration of these services' unit files, that we're not checking or validating
+# anywhere, so we need to define this behaviour more precisely by providing additional tests.
+Watchdog doesn't send notification if service is down which leads to service restart
+    Set Service health check response    response=0
+    Restart Service    ${SERVICE_NAME}
+    Restart Service    tedge-watchdog
+
+    # wait for service to start
+    ${pid} =    Service Should Be Running    ${SERVICE_NAME}
+
+    # send status up
+    ${ts} =    Get Unix Timestamp
+
+    # wait for a health check command from watchdog
+    Should Have MQTT Messages    te/device/main/service/${SERVICE_NAME}/cmd/health/check
+
+    Sleep    ${WATCHDOG_SEC}s
+
+    # Verify that service got restarted by systemd and it's now healthy
+    ${pid1} =    Service Should Be Running    ${SERVICE_NAME}
+    Should Not Be Equal    ${pid}    ${pid1}
+
+Watchdog doesn't fail on unexpected time format
+    Set Service health check response    response=1
+    Restart Service    ${SERVICE_NAME}
+    Restart Service    tedge-watchdog
+
+    # wait for service to start
+    ${pid} =    Service Should Be Running    tedge-watchdog
+
+    # send status up
+    ${ts} =    Get Unix Timestamp
+    Execute Command
+    ...    tedge mqtt pub 'te/device/main/service/${SERVICE_NAME}/status/health' '{"status": "down", "pid": ${pid}, "time": ${ts}}'
+
+    # wait for a health check command from watchdog
+    Should Have MQTT Messages    te/device/main/service/${SERVICE_NAME}/cmd/health/check
+
+    # Simulate service signalling that it's unhealthy with wrong format in `time` field
+    # (we're expecting a UNIX timestamp to be an int, not a string)
+    ${before_restart_timestamp} =    Get Unix Timestamp
+    Execute Command
+    ...    tedge mqtt pub 'te/device/main/service/${SERVICE_NAME}/status/health' '{"status": "down", "pid": ${pid}, "time": "${before_restart_timestamp}"}'
+
+    # Wait until service is restarted due to tedge-watchdog not sending out the notification
+    Sleep    ${WATCHDOG_SEC}s
+
+    # Verify that tedge-watchdog is still running
+    ${pid1} =    Service Should Be Running    tedge-watchdog
+    Should Be Equal    ${pid}    ${pid1}
```

Comment:
Shouldn't we also validate that the actual service itself is also restarted, when it keeps sending bad responses?

Reply:
Indeed, apart from checking that tedge-watchdog doesn't crash, we can also check if it still functions correctly.
Also the test was buggy: tedge-watchdog PID was used in a message from a service, which was fixed.
fixed cdca9120e4